// @nevware21/ts-async/dist/es6/mod/promise/asyncPromise.js
import { _createAllPromise, _createAllSettledPromise, _createAnyPromise, _createPromise, _createRacePromise, _createRejectedPromise, _createResolvedPromise } from "./base";
import { timeoutItemProcessor } from "./itemProcessor";
let _allAsyncSettledCreator;
let _raceAsyncCreator;
let _anyAsyncCreator;
export function createAsyncPromise(executor, timeout) {
    return _createPromise(createAsyncPromise, timeoutItemProcessor(timeout), executor, timeout);
}
export const createAsyncAllPromise = _createAllPromise(createAsyncPromise);
export const createAsyncResolvedPromise = _createResolvedPromise(createAsyncPromise);
export const createAsyncRejectedPromise = _createRejectedPromise(createAsyncPromise);
export function createAsyncAllSettledPromise(input, timeout) {
    !_allAsyncSettledCreator && (_allAsyncSettledCreator = _createAllSettledPromise(createAsyncPromise));
    return _allAsyncSettledCreator.v(input, timeout);
}
export function createAsyncRacePromise(values, timeout) {
    !_raceAsyncCreator && (_raceAsyncCreator = _createRacePromise(createAsyncPromise));
    return _raceAsyncCreator.v(values, timeout);
}
export function createAsyncAnyPromise(values, timeout) {
    !_anyAsyncCreator && (_anyAsyncCreator = _createAnyPromise(createAsyncPromise));
    return _anyAsyncCreator.v(values, timeout);
}
