// @nevware21/ts-async/dist/es6/mod/promise/idlePromise.js
import { isUndefined } from "@nevware21/ts-utils";
import { _createAllPromise, _createAllSettledPromise, _createAnyPromise, _createPromise, _createRacePromise, _createRejectedPromise, _createResolvedPromise } from "./base";
import { idleItemProcessor } from "./itemProcessor";
let _defaultIdleTimeout;
let _allIdleSettledCreator;
let _raceIdleCreator;
let _anyIdleCreator;
export function setDefaultIdlePromiseTimeout(idleDeadline) {
    _defaultIdleTimeout = idleDeadline;
}
export const setDefaultIdleTimeout = (setDefaultIdlePromiseTimeout);
export function createIdlePromise(executor, timeout) {
    let theTimeout = isUndefined(timeout) ? _defaultIdleTimeout : timeout;
    return _createPromise(createIdlePromise, idleItemProcessor(theTimeout), executor, theTimeout);
}
export const createIdleAllPromise = _createAllPromise(createIdlePromise);
export const createIdleResolvedPromise = _createResolvedPromise(createIdlePromise);
export const createIdleRejectedPromise = _createRejectedPromise(createIdlePromise);
export function createIdleAllSettledPromise(input, timeout) {
    !_allIdleSettledCreator && (_allIdleSettledCreator = _createAllSettledPromise(createIdlePromise));
    return _allIdleSettledCreator.v(input, timeout);
}
export function createIdleRacePromise(values, timeout) {
    !_raceIdleCreator && (_raceIdleCreator = _createRacePromise(createIdlePromise));
    return _raceIdleCreator.v(values, timeout);
}
export function createIdleAnyPromise(values, timeout) {
    !_anyIdleCreator && (_anyIdleCreator = _createAnyPromise(createIdlePromise));
    return _anyIdleCreator.v(values, timeout);
}
