// @nevware21/ts-async/dist/es6/mod/promise/syncPromise.js
import { _createAllPromise, _createAllSettledPromise, _createAnyPromise, _createPromise, _createRacePromise, _createRejectedPromise, _createResolvedPromise } from "./base";
import { syncItemProcessor } from "./itemProcessor";
let _allSyncSettledCreator;
let _raceSyncCreator;
let _anySyncCreator;
export function createSyncPromise(executor) {
    return _createPromise(createSyncPromise, syncItemProcessor, executor);
}
export const createSyncAllPromise = _createAllPromise(createSyncPromise);
export const createSyncResolvedPromise = _createResolvedPromise(createSyncPromise);
export const createSyncRejectedPromise = _createRejectedPromise(createSyncPromise);
export function createSyncAllSettledPromise(input, timeout) {
    !_allSyncSettledCreator && (_allSyncSettledCreator = _createAllSettledPromise(createSyncPromise));
    return _allSyncSettledCreator.v(input, timeout);
}
export function createSyncRacePromise(values, timeout) {
    !_raceSyncCreator && (_raceSyncCreator = _createRacePromise(createSyncPromise));
    return _raceSyncCreator.v(values, timeout);
}
export function createSyncAnyPromise(values, timeout) {
    !_anySyncCreator && (_anySyncCreator = _createAnyPromise(createSyncPromise));
    return _anySyncCreator.v(values, timeout);
}
